subsys/bluetooth/controller/ll_sw/ull_conn_iso_types.h

```diff
@@ -48,6 +48,10 @@ struct ll_conn_iso_stream {
 				    */
 	uint8_t  terminate_reason;
 	uint8_t  cis_id;
+
+#if defined(CONFIG_BT_CTLR_ISOAL_PSN_IGNORE)
+	uint64_t pkt_seq_num:39;
+#endif /* CONFIG_BT_CTLR_ISOAL_PSN_IGNORE */
```

Comment:
Why do we only include the PSN if we ignore it?

Reply:
When PSN from HCI is ignored, ISOAL still needs to schedule it to successive ISO events, and this `pkt_seq_num` value is used to generate the PSN so that ISOAL implementation is not modified when ignoring PSN from HCI.

i.e. CONFIG_BT_CTLR_ISOAL_PSN_IGNORE implementation is ensuring no HCI ISO data is considered stale by ISOAL implementation.

Reply:
So if the PSN from HCI is ignore, you are generating fake PSNs in the controller?

I see that the code seems to just be incrementing the PSN (e.g. `cis->pkt_seq_num++;`), but does that work properly if the host does not send an SDU per SDU interval? 

If ignoring PSN from HCI, shouldn't the packets just be put into some kind of FIFO and the ISOAL not care about the PSN at all? 

Finally, how does the timestamp come into picture here? If you ignore the PSN and the host sends an SDU with timestamp = 10000 and then afterwards an SDU with timestamp = 9000, then the latter SDU should be sent first, right?

Reply:
> If ignoring PSN from HCI, shouldn't the packets just be put into some kind of FIFO and the ISOAL not care about the PSN at all?

ISOAL simply assigns a payload count that will be picked by the lower link layer, it does not maintain or check next sequence number. Lower link layer check the assigned payload count and decides to drop ISO PDU with stale payload count. Hence, to truely enqueue as a simple FIFO, fake sequence of sequence number that is not stale compared to current lower link layer event count has to be given to ISOAL.

> Finally, how does the timestamp come into picture here? If you ignore the PSN and the host sends an SDU with timestamp = 10000 and then afterwards an SDU with timestamp = 9000, then the latter SDU should be sent first, right?

PSN_IGNORE should not be enabled if host intended to use timestamp, and its clock is in sync with the controller clock. ISOAL will give a correct payload count for timestamp = 10000 and will give the same last payload count for SDU with timestamp = 9000. Later SDU will be dropped by the lower link layer as stale.

Reply:
Alright, so PSN_IGNORE effectively also disables the use of timestamps from the host? 

When the PSN_IGNORE feature is disable, is the controller able to handle the case where the host sends the following at t=0 
SDU 1 with PSN 5 and timestamp 5000
SDU 2 with PSN 10 and timestamp 10000
SDU 3 with PSN 9 and timestamp 9000

Or will it drop SDU 3?

Reply:
The implementation states this:

https://github.com/zephyrproject-rtos/zephyr/blob/1dc6a8aa13fa73f9276e411d46d72d983ede0487/subsys/bluetooth/controller/ll_sw/isoal.c#L1837-L1842

Hence in the case of your example, 5, 10 and 9, are handled such that they are enqueued towards LLL in same order with payload counts 5, 10 and 10, where SDU 3 will be dropped in the LLL.

Reply:
I see, so packets are dropped on PSN based on the order they come in, rather than the actual value of it? So the host sending PSN 9 after PSN 10 will always get dropped?

Reply:
@niag-Oticon could you confirm my understanding that `isoal.c` enqueues ISO SDU with same payload number/count when the out-of-order PSN is provided by Host?

@Thalley LLL discards/drops ISO PDU enqueued by ISOAL if they do not have correct order payload_count/number assigned to them (by ISOAL) corresponding to ISO SDU PSN value.

Reply:
I don't fully understand the implementation of the fake packet sequence number. However to address the out of order SDUs focusing only on the ISO-AL implementation:
SDU 1 with PSN 5 and timestamp 5000
SDU 2 with PSN 10 and timestamp 10000
SDU 3 with PSN 9 and timestamp 9000

The ISO-AL was not designed for this. Underlying assumption is that this is not a realistic case according to specifications.

But if it was to happen with CONFIG_BT_CTLR_ISOAL_SN_STRICT enabled, given the present logic (Unframed):

SDU 1 at PSN 5 TS 5000      -----> PDU 5 Event 5 (Assuming that target event is 5 on input to ISO-AL)
SDU 2 at PSN 10 TS 10000  ------> PDU 10 Event 10
SDU 3 at PSN 9 TS 9000      ------> PDU 11 Event 11

If it was to happen with CONFIG_BT_CTLR_ISOAL_SN_STRICT enabled, given the present logic (Framed):

SDU 1 at PSN 5 TS 5000      -----> PDU 5 Offset 0 Event 5 (Assuming that target event is 5 on input to ISO-AL)
SDU 2 at PSN 10 TS 10000  ------> PDU 10 Offset 0 Event 10
SDU 3 at PSN 9 TS 9000      ------> PDU 11 Offset 2000 Event 11

Reply:
@niag-Oticon if there is no concern or impact or issue on your platform, can i get a review/approval for the ULL/HCI changes?